diff --git a/core/jni/Android.mk b/core/jni/Android.mk
index b5a2f98..82c46a3 100644
--- a/core/jni/Android.mk
+++ b/core/jni/Android.mk
@@ -23,6 +23,10 @@ ifeq ($(USE_OPENGL_RENDERER),true)
 	LOCAL_CFLAGS += -DUSE_OPENGL_RENDERER
 endif
 
+ifeq($(TARGET_USES_ICS_CAMERA_BLOB),true)
+	LOCAL_CFLAGS += -DICS_CAMERA_BLOB
+endif
+
 LOCAL_CFLAGS += -DGL_GLEXT_PROTOTYPES -DEGL_EGLEXT_PROTOTYPES
 
 LOCAL_SRC_FILES:= \
diff --git a/core/jni/android_hardware_Camera.cpp b/core/jni/android_hardware_Camera.cpp
index 6cd8955..9dab1d2 100644
--- a/core/jni/android_hardware_Camera.cpp
+++ b/core/jni/android_hardware_Camera.cpp
@@ -813,12 +813,16 @@ static void android_hardware_Camera_stopFaceDetection(JNIEnv *env, jobject thiz)
 static void android_hardware_Camera_enableFocusMoveCallback(JNIEnv *env, jobject thiz, jint enable)
 {
     ALOGV("enableFocusMoveCallback");
+#ifdef ICS_CAMERA_BLOB
+    return;
+#else
     sp<Camera> camera = get_native_camera(env, thiz, NULL);
     if (camera == 0) return;
 
     if (camera->sendCommand(CAMERA_CMD_ENABLE_FOCUS_MOVE_MSG, enable, 0) != NO_ERROR) {
         jniThrowRuntimeException(env, "enable focus move callback failed");
     }
+#endif
 }
 
 //-------------------------------------------------
